okhttp/src/test/java/io/grpc/okhttp/AsyncFrameWriterTest.java

```diff
@@ -0,0 +1,138 @@
+import static org.mockito.Mockito.verify;
+import static org.mockito.internal.verification.VerificationModeFactory.times;
+
+@RunWith(MockitoJUnitRunner.class)
+public class AsyncFrameWriterTest {
+
+  @Mock private OkHttpClientTransport transport;
+  @Mock private Socket socket;
+  @Mock(answer = Answers.CALLS_REAL_METHODS)
+  private PartiallyTrackingFrameWriter frameWriter;
+
+  private QueueingExecutor queueingExecutor;
+  private AsyncFrameWriter asyncFrameWriter;
+
+  @Before
+  public void setUp() throws Exception {
+    queueingExecutor = new QueueingExecutor();
+    asyncFrameWriter = spy(new AsyncFrameWriter(transport, new SerializingExecutor(queueingExecutor)));
+    asyncFrameWriter.becomeConnected(frameWriter, socket);
+  }
+
+  @After
+  public void tearDown() throws Exception {
+    asyncFrameWriter.close();
+  }
+
+  @Test
+  public void noCoalesceRequired() throws IOException {
+    asyncFrameWriter.ping(true, 0, 1);
+    asyncFrameWriter.flush();
+    queueingExecutor.runAll();
+
+    verify(frameWriter, times(1)).ping(anyBoolean(), anyInt(), anyInt());
+    verify(asyncFrameWriter, times(1)).flush();
+    verify(frameWriter, times(1)).flush();
+  }
+
+  @Test
+  public void flushCoalescing_shouldNotMergeTwoDistinctFlushes() throws IOException {
+    asyncFrameWriter.ping(true, 0, 1);
+    asyncFrameWriter.flush();
+    queueingExecutor.runAll();
+
+    asyncFrameWriter.ping(true, 0, 2);
+    asyncFrameWriter.flush();
+    queueingExecutor.runAll();
+
+    verify(frameWriter, times(2)).ping(anyBoolean(), anyInt(), anyInt());
+    verify(asyncFrameWriter, times(2)).flush();
+    verify(frameWriter, times(2)).flush();
+  }
+
+  @Test
+  public void flushCoalescing_shouldMergeTwoQueuedFlushes() throws IOException {
+    asyncFrameWriter.ping(true, 0, 1);
+    asyncFrameWriter.flush();
+    asyncFrameWriter.ping(true, 0, 2);
+    asyncFrameWriter.flush();
+
+    queueingExecutor.runAll();
+
+    verify(frameWriter, times(2)).ping(anyBoolean(), anyInt(), anyInt());
+    verify(asyncFrameWriter, times(2)).flush();
+    verify(frameWriter, times(1)).flush();
+    assertThat(Iterables.getLast(PartiallyTrackingFrameWriter.getAllMethodsInOrder()))
+        .isSameAs(PartiallyTrackingFrameWriter.Method.FLUSH);
+  }
+
+  /**
+   * Executor queues incoming runnables instead of running it. Runnables can be invoked via {@code
+   * runAll} in serial order.
+   */
+  private static class QueueingExecutor implements Executor {
+
+    private final Queue<Runnable> runnables = new ConcurrentLinkedQueue<Runnable>();
+
+    @Override
+    public void execute(Runnable command) {
+      runnables.add(command);
+    }
+
+    public void runAll() {
+      Runnable r;
+      while ((r = runnables.poll()) != null) {
+        r.run();
+      }
+    }
+  }
+
+  private static abstract class PartiallyTrackingFrameWriter implements FrameWriter {
+
+    private static final List<Method> methods = new ArrayList<Method>();
```

Comment:
This is just asking for trouble, as one test can influence another. Actually... how is flushCoalescing_shouldMergeTwoQueuedFlushes working? Nothing is clearing out these methods.